mutagen/wave.py

```diff
@@ -0,0 +1,369 @@
+from mutagen._riff import RiffFile, RiffChunkHeader
+from mutagen.id3._util import ID3NoHeaderError, error as ID3Error
+from mutagen._util import resize_bytes, delete_bytes, MutagenError, loadfile, \
+    convert_error
+
+__all__ = ["WAVE", "Open", "delete"]
+
+
+class error(MutagenError):
+    pass
+
+
+class InvalidChunk(error):
+    pass
+
+
+def is_valid_chunk_id(id):
+    # looks like this is failing if python is not started with -bb in TravisCI:
+    # assert isinstance(id, text_type)
+
+    return ((len(id) <= 4) and (min(id) >= u' ') and
+            (max(id) <= u'~'))
+
+
+def assert_valid_chunk_id(id):
+    if not is_valid_chunk_id(id):
+        raise KeyError("RIFF/WAVE-chunk-Id must be four ASCII characters.")
+
+
+class WaveChunk(object):
+    """Representation of a common WaveChunk"""
+
+    #  Chunk headers are 8 bytes long (4 for ID and 4 for the size)
+    HEADER_SIZE = 8
+
+    def __init__(self, fileobj, parent_chunk=None):
+        self.__fileobj = fileobj
+        self.parent_chunk = parent_chunk
+        self.offset = fileobj.tell()
+
+        header = fileobj.read(self.HEADER_SIZE)
+        if len(header) < self.HEADER_SIZE:
+            raise InvalidChunk()
+
+        self.id, self.data_size = struct.unpack('<4sI', header)
+
+        try:
+            self.id = self.id.decode('ascii')
+        except UnicodeDecodeError:
+            raise InvalidChunk()
+
+        if not is_valid_chunk_id(self.id):
+            raise InvalidChunk()
+
+        self.size = self.HEADER_SIZE + self.data_size
+        self.data_offset = fileobj.tell()
+
+    def read(self):
+        """Read the chunks data"""
+
+        self.__fileobj.seek(self.data_offset)
+        return self.__fileobj.read(self.data_size)
+
+    def write(self, data):
+        """Write the chunk data"""
+
+        if len(data) > self.data_size:
+            raise ValueError
+
+        self.__fileobj.seek(self.data_offset)
+        self.__fileobj.write(data)
+
+    def delete(self):
+        """Removes the chunk from the file"""
+
+        delete_bytes(self.__fileobj, self.size, self.offset)
+        if self.parent_chunk is not None:
+            self.parent_chunk._update_size(
+                self.parent_chunk.data_size - self.size)
+
+    def _update_size(self, data_size):
+        """Update the size of the chunk"""
+
+        self.__fileobj.seek(self.offset + 4)
+        self.__fileobj.write(pack('>I', data_size))
+        if self.parent_chunk is not None:
+            size_diff = self.data_size - data_size
+            self.parent_chunk._update_size(
+                self.parent_chunk.data_size - size_diff)
+        self.data_size = data_size
+        self.size = data_size + self.HEADER_SIZE
+
+    def resize(self, new_data_size):
+        """Resize the file and update the chunk sizes"""
+
+        resize_bytes(
+            self.__fileobj, self.data_size, new_data_size, self.data_offset)
+        self._update_size(new_data_size)
+
+
+class WaveFile(RiffFile):
```

Comment:
I agree it is messy.

I started of with a copy of aiff.py because I knew it had a pretty simular structure.

The idea is to move everything what is part of the RIFF standard into _riff.py, and WAVE specific stuff into wave.py. Which was easier said then done, because is was not always clear to me what belonged to core RIFF and what to MS-WAVE. Because the sub-chunks used a little endian size, in contradiction to the RIFF chuck using big endian Therefor I though the sub-chunk definition are of sub standard, which does not seem to be the case after all. I found somewhere that chunks with capital tags are shared chunks and lower case type specific. 

Anyway, I will do my best do clean up the mess.